### PR TITLE
Enhancement for container prune and volume prune(issue #648)

### PR DESCRIPTION
**Pull request by click2cloud-lamda**
Issue Fixed: #648 

Enhancements for container prune and volume prune. Only stopped and created containers are deleted and only unmounted volumes are deleted.

```
D:\GolandProjects\github.com\nerdctl\cmd\nerdctl>nerdctl container prune
WARNING! This will remove all stopped containers.
Are you sure you want to continue? [y/N] y



Deleted Containers
001251e3770003c5918a81513f7073377ba6a14089b607878fe82c377e097738
72fe18497ae4f9b0e42dc3c1fee236de781c100142b073385847d53871abd16b
7bdee6ba154ae1c7eb05201ff0deed21db16d1ade978a310b9a079e15a71b669

```

```
D:\GolandProjects\github.com\nerdctl\cmd\nerdctl>nerdctl volume prune
WARNING! This will remove all local volumes not used by at least one container.
Are you sure you want to continue? [y/N] y
Deleted Volumes
testVolume123
testVolume
testVolume007
```

Signed-off-by: Raymond Mathew <raymond.mathew@click2cloud.net>

### CODE REVIEW

File: cmd/nerdctl/volume_prune.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		if err != nil {
+			return err
+		}
+		fmt.Fprintln(cmd.OutOrStdout(), "Deleted Volumes")
+		print(rNames, cmd)
+	}
+	return nil
+}
+
+func print(x []string, cmd *cobra.Command) {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/volume_prune.go:109]: This conflicts with built-in `print`.
Please come up with a unique and descriptive name.

File: cmd/nerdctl/container_prune.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	containerPruneCommand := &cobra.Command{
+		Use:           "prune [flags]",
+		Short:         "Remove all stopped containers",
+		RunE:          containerPruneAction,
+		SilenceUsage:  true,
+		SilenceErrors: true,
+	}
+	containerPruneCommand.Flags().BoolP("force", "f", false, "Do not prompt for confirmation")
+	return containerPruneCommand
+}
```
Comment (AkihiroSuda) [inline cmd/nerdctl/container_prune.go:40]: Could you update the Command Reference in `README.md` too?

File: cmd/nerdctl/volume_prune.go (hunk 1 above)
Comment (click2cloud-lamda) [reply cmd/nerdctl/volume_prune.go:109]: Okay

File: cmd/nerdctl/container_prune.go (hunk 2 above)
Comment (click2cloud-lamda) [reply cmd/nerdctl/container_prune.go:40]: Sure!

File: cmd/nerdctl/container_prune.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	ctx = namespaces.WithNamespace(ctx, ns)
+	var deletedContainers []string
+	containers, err := client.Containers(ctx)
+	if err != nil {
+		return err
+	}
+	for _, container := range containers {
+		cont, _ := containerinspector.Inspect(ctx, container)
```
Comment (Junnplus) [inline cmd/nerdctl/container_prune.go:77]: Check status by task.

**Review by Junnplus — CHANGES_REQUESTED** [review]

**Comment by Junnplus** [comment]
Duplicate of https://github.com/containerd/nerdctl/pull/806 ?

**Comment by click2cloud-lamda** [comment]
> Duplicate of #806 ?

The container_prune code of #806 doesnt delete the containers which are just created and not started.

File: cmd/nerdctl/volume_prune.go (hunk 1 above)
Comment (AkihiroSuda) [reply cmd/nerdctl/volume_prune.go]: > Display

"Display" isn't unique/descriptive, and I still don't get what this function does and why this function is needed.

Comment (click2cloud-lamda) [reply cmd/nerdctl/volume_prune.go]: This function displays the containers one-by-one in this way:
```
Deleted Volumes
testVolume123
testVolume
testVolume007
```
and not in this way:
```
Deleted Volumes: [testVolume123 testVolume testVolume007]
```

File: README.md [hunk 4]
```diff
@@ -660,6 +662,14 @@ Flags:
 
 Unimplemented `docker inspect` flags:  `--size`
 
+### :whale: :blue_square: nerdctl container prune
+Removes all stopped containers
+
+Usage: `nerdctl container prune [OPTIONS] NAME|ID [NAME|ID...]`
+
+Flags:
+- :whale: `-f, --force`: Do not prompt for confirmation
```
[the remark this answers is not in the document]
Comment (click2cloud-lamda) [reply README.md:671]: Okay

File: cmd/nerdctl/container_prune.go (hunk 3 above)
Comment (click2cloud-lamda) [reply cmd/nerdctl/container_prune.go:78]: If I check with task, it will throw errors for containers without tasks.

Comment (junnplus) [reply cmd/nerdctl/container_prune.go:78]: You may need `errdefs.IsNotFound`

Comment (junnplus) [reply cmd/nerdctl/container_prune.go:78]: In docker, if the status is created, it seems to be remove.

Comment (click2cloud-lamda) [reply cmd/nerdctl/container_prune.go:78]: > In docker, if the status is created, it seems to be remove.

It also does the same. Removes the created containers as well.

**Comment by click2cloud-lamda** [comment]
> 

You mean creating volumes and containers and removing them using prune.

**Comment by click2cloud-lamda** [comment]
> Can we have an integration test?
You mean creating volumes and containers and removing them using prune.

**Comment by click2cloud-lamda** [comment]
> Can we have an integration test?

You mean creating volumes and containers and removing them using prune.

**Comment by junnplus** [comment]
> You mean creating volumes and containers and removing them using prune.

Yes.

File: cmd/nerdctl/container_prune.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		if strings.ToLower(confirm) != "y" {
+			return nil
+		}
+	}
+	ns, err := cmd.Flags().GetString("namespace")
+	if err != nil {
+		return err
+	}
+
+	ctx = namespaces.WithNamespace(ctx, ns)
```
Comment (junnplus) [inline cmd/nerdctl/container_prune.go:71]: we don't need it, see https://github.com/containerd/nerdctl/blob/master/cmd/nerdctl/client.go#L44

File: cmd/nerdctl/container_prune.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	var deletedContainers []string
+	containers, err := client.Containers(ctx)
```
Comment (junnplus) [inline cmd/nerdctl/container_prune.go:73]: Can we filter the containers that need to be deleted?

**Review by junnplus — CHANGES_REQUESTED** [review]

File: cmd/nerdctl/container_prune.go (hunk 5 above)
Comment (click2cloud-lamda) [reply cmd/nerdctl/container_prune.go:71]: Okay

File: cmd/nerdctl/container_prune.go (hunk 6 above)
Comment (click2cloud-lamda) [reply cmd/nerdctl/container_prune.go:73]: Filter on what basis?

Comment (click2cloud-lamda) [reply cmd/nerdctl/container_prune.go:73]: Let me work on it.

Comment (click2cloud-lamda) [reply cmd/nerdctl/container_prune.go:82]: I have added filters for volume and container according to docker

**Pull request closed without merging**